tap_dynamodb/sync_strategies/log_based.py

```diff
@@ -174,63 +163,69 @@ def sync_log_based(config, state, stream):
-                seq_number_bookmarks.pop(shard['ShardId'])
-                finished_shard_bookmarks.append(shard['ShardId'])
-                state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
-                state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
-                singer.write_state(state)
-
-    for shard in finished_shard_bookmarks:
-        if shard not in found_shards:
+        # Now that we have fully synced the shard, move it from the
+        # shard_seq_numbres to finished_shards. Must check if the bookmark
+        # exists because if a shard has 0 records we will never set a
+        # bookmark for the shard
+        if seq_number_bookmarks.get(shard['ShardId']):
+            seq_number_bookmarks.pop(shard['ShardId'])
+            finished_shard_bookmarks.append(shard['ShardId'])
+            state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
+            state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
+            singer.write_state(state)
+
+    for shardId in finished_shard_bookmarks:
+        if shardId not in found_shards:
             # Remove this shard because its no longer appearing when we query for get_shards
-            finished_shard_bookmarks.remove(shard)
+            finished_shard_bookmarks.remove(shardId)
             state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
 
     singer.write_state(state)
 
     return rows_synced
 
-def has_stream_aged_out(config, state, stream):
+
+def has_stream_aged_out(state, table_name):
+    '''
+    Uses the success_timestamp on the stream to determine if we have
+    successfully synced the stream in the last 19 hours 30 minutes.
+
+    This uses 19 hours and 30 minutes because records are guaranteed to be
+    available on a shard for 24 hours, but we only process closed shards
+    and shards close after 4 hours. 24-4 = 20 and we gave 30 minutes of
+    wiggle room because I don't trust AWS.
+    '''
+    current_time = singer.utils.now()
+
+    success_timestamp = singer.get_bookmark(state, table_name, 'success_timestamp')
+
+    # If we have no success_timestamp then we have aged out
+    if not success_timestamp:
+        return True
+
+    time_span = current_time - singer.utils.strptime_to_utc(success_timestamp)
+
+    # If it has been > than 19h30m since the last successful sync of this
+    # stream then we feel confident we have not aged out
+    return time_span > datetime.timedelta(hours=19, minutes=30)
+
+
+def get_initial_bookmarks(config, state, table_name):
+    '''
+    Returns the state including all bookmarks necessary for the initial
+    full table sync
+    '''
+    # TODO This can be improved by getting the max sequence number in the
+    # open shards and include those in the seq_number bookmark
```

Comment:
@cosimon I'm happy to implement this TODO if you think it's worth doing. I think for the. initial bookmarking it's a good idea personally. It would reduce the number of updates / changes after a full table sync but before the first time we start getting stream data from shards open at the time we're doing full table sync.

Reply:
Discussed in slack, will create a backlog ticket.